Articles/FileSystemProviders/extending.md

```diff
@@ -0,0 +1,213 @@
+
+    public ImagesController(MediaFileManager mediaFileManager)
+    {
+        _mediaFileManager = mediaFileManager;
+    }
+
+{...}
+```
+
+You can then access the configured file system provider through `_mediaFileManager.FileSystem`, which is the same way Umbraco will access the file system provider.
+
+## MediaPath Scheme
+
+The MediaPath Scheme defines the current set of rules that decide the format of the Media Path when it is saved into the media archive wherever it is located.
+
+By default the MediaPath scheme used by Umbraco is the `UniqueMediaPathScheme` this generates a unique 'folder' to place the uploaded image in eg.
+
+`/media/dozdrg2f/mylovelyimage.jpg`
+
+`/media` is defined by the PhysicalFileSystem and 'dozdrg2f' is generated by the `UniqueMediaPathScheme`.
+
+:::note
+In Umbraco 7 the integer ids were used in the path, and this approach is still possible using the 'OriginalMediaPathScheme'
+:::
+
+You can set the `MediaPathScheme` during composition, for example if you wanted to revert back to the V7 methodology in a migrated site:
+
+```c#
+  builder.Services.AddUnique<IMediaPathScheme, OriginalMediaPathScheme>();
+```
+
+And you could create your own logic for the path by implementing `IMediaPathScheme`
+
+## Other IFileSystems
+
+Umbraco also registers instances of PhysicalFileSystem for the following parts of Umbraco that persist to 'files':
+
+- MacroPartialsFileSystem
+- PartialViewsFileSystem
+- StylesheetsFileSystem
+- ScriptsFileSystem
+- MvcViewsFileSystem 
+
+These are accessible via dependency injection
+
+```csharp
+public class FileSystemLocations 
+{
+    private readonly FileSystems _fileSystems;
+    public FileSystemLocations(FileSystems fileSystems)
+    {
+        _fileSystems = fileSystems;
+        var macroPartialsFileSystem = _fileSystems.MacroPartialsFileSystem;
+    }     
+```
+
+`IFileSystem`, `MediaFileManager`, and `FileSystems` are located in the `Umbraco.Cms.Core.IO` namespace.
+
+### Stylesheet Filesystem
+
+Like with the media file system it is also possible to replace the stylesheet filesystem with your own implementation of `IFileSystem` in a composer. It's important to note here that, unlike media file system, you cannot replace the filesystem with a `PhysicalFileSystem` using a different root path or root URL, this will not work, and will cause issues since the root path is coupled to the virtual path, given by the frontend, e.g. `/css/MyBeautifulStyle.css`. 
+
+When replacing the stylesheet filesystem, you don't need to register it, since it's only available through Filesystems, what you need to do instead is configure the `FileSystems` to use your implementation for the StylesheetsFileSystem.
+
+The IUmbracoBuilder has an extension method for configuring the `FileSystems`, you need to invoke this method with an action that accepts an `IServiceProvider` and the `FileSystems` you will configure, configuring the `FileSystems` can look like this:
```

Comment:
```suggestion
Like with the media file system it is also possible to replace the stylesheet filesystem with your own implementation of `IFileSystem` in a composer. It's important to note here that, unlike media file system, you cannot replace the filesystem with a `PhysicalFileSystem` using a different root path or root URL, this will not work, and will cause issues since the root path is coupled to the virtual path, given by the frontend, e.g. `/css/MyBeautifulStyle.css`. 

When replacing the stylesheet filesystem, you don't need to register it, since it's only available through Filesystems, what you need to do instead is configure the `FileSystems` to use your implementation for the `StylesheetsFileSystem`.

The IUmbracoBuilder has an extension method for configuring the `FileSystems`, you need to invoke this method with an action that accepts an `IServiceProvider` and the `FileSystems` you will configure, configuring the `FileSystems` can look like this:
```